lib/vsc/mympirun/pmi/slurm.py

```diff
@@ -0,0 +1,260 @@
+# along with vsc-mympirun.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+Slurm PMI class, i.e. wrap around srun
+"""
+
+import os
+import re
+
+from vsc.mympirun.pmi.sched import Sched
+from vsc.mympirun.pmi.pmi import PMI, PMIX
+from vsc.utils.run import async_to_stdout
+from vsc.mympirun.pmi.option import DISTRIBUTE_PACK, DISTRIBUTE_CYCLE
+
+DISTRIBUTE_MAP = {
+    DISTRIBUTE_PACK: 'block:block:block',
+    DISTRIBUTE_CYCLE: 'cycle:cycle:cycle',
+}
+
+
+class Slurm(Sched):
+
+    _sched_for = ['slurm']
+    SCHED_ENVIRON_ID = 'SLURM_JOB_ID'
+
+    LAUNCHER = 'srun'
+
+    def run_function(self):
+        """
+        srun itself supports redirection to file
+        """
+        args = []
+        if self.options.output:
+            args = ['--output=%s' % self.options.output]
+
+        return async_to_stdout, args
+
+    def pmicmd_sched(self):
+        """Generate the sched related arguments to the launcher as a list"""
+        args = []
+
+        # sbatch will set the sbatch parameters as environment variables and they will get picked up by srun
+        args.append("--chdir=" + os.getcwd())
+
+        if self.options.distribute:
+            args.append("--distribution=%s" % DISTRIBUTE_MAP[self.options.distribute])
+
+        return args
+
+    def pmicmd_environment(self):
+        """export everything"""
+        return ['--export=ALL']
+
+    def pmicmd_mpi(self):
+        """Handle PMI to --mpi"""
+        for pmi in self.PMI:
+            flavour = None
+            if pmi.FLAVOUR == PMIX:
+                flavour = 'pmix'
+                if pmi.VERSION == 3:
+                    flavour += '_v3'
+                else:
+                    self.log.warn("Unsupported PMIx version %s, trying with generic %s", pmi, flavour)
+            elif pmi.FLAVOUR == PMI:
+                if pmi.VERSION == 2:
+                    flavour = 'pmi2'
+                else:
+                    self.log.error("Unsupported PMI version %s", pmi)
+            else:
+                self.log.error("Unsupported PMI %s" % pmi)
+
+            if flavour is not None:
+                # first one wins
+                self.log.debug("Mapped PMI %s to flavour %s", pmi, flavour)
+                return ["--mpi=%s" % flavour]
+
+        self.log.error("No supported PMIs")
+        return []
+
+    def _job_info(self, job_info):
+        """
+        Fill in/complete/edit job_info dict and return it
+
+        # relevant variables to use
+        SLURM_CPUS_ON_NODE=32
+        SLURM_JOB_CPUS_PER_NODE=32(x2)
+        SLURM_JOB_GPUS=0,1,2,3
+        SLURM_JOB_NODELIST=node[3302-3303]
+        SLURM_JOB_NUM_NODES=2
+        SLURM_MEM_PER_CPU=7600
+        SLURM_NNODES=2
+        SLURM_NPROCS=64
+        SLURM_NTASKS=64
+        """
+        se = {}  # slurm enviroment
+
+        # TODO: check here if the requested gpus are usable or not (eg corresponding CUDA_DEVICES)
+
+        dbg = []
+        prefix = 'SLURM_'
+        for k, v in os.environ.items():
```

Comment:
Haven't we learned not to use single-letter variables outside of list comprehensions? Use `key, val`?

Reply:
clearly not :wink: